### ISSUE TITLE
[Feature]: Update Command + Update on Shutdown flag

### ISSUE THREAD

**Issue by mbround18** [mbround18/valheim-docker#129]
> Hi,
> 
> Thank you kindly for your work on this, really great docker image and it has been working 
> well for me. I do have a suggestion for the auto update script though - if there's no update 
> available, don't shut down the server. This would allow for more frequent update checks as 
> well, instead of just once per day (For those that would rather be updated asap).
> I believe this is possible via SteamCMD, though it's a bit of a hacky workaround (Skip to the last two posts in this thread): 
> https://steamcommunity.com/app/346110/discussions/0/535152511358957700/
> If I were more familiar with rust I'd offer up a PR but alas it's all a bit greek to me.

_Originally posted by @neoKushan in https://github.com/mbround18/valheim-docker/discussions/126_

Acceptance Criteria: 
- Odin Update command which checks for updates
- Environment Flag to turn on checking for updates upon shutdown

**Comment by FetaCini** [comment]
In addition to this it would be great if it could do a check before it shuts down to see if anyone is currently in the server and wait till user count = 0 before shutting down and updating.

**Comment by LovecraftianHorror** [comment]
Alright, so I'm looking through the full thread currently. There's a lot of information from quite a while ago, so I'm planning on spinning up a valheim server with no auto updating so that it can go stale after the next update and I can use it for testing to see if any of the information is out of date.

_Note: to keep things easy to digest I will refer to the duplicate server install used to check for updates as "staging"_

Currently it looks like this method _might_ involve having to keep a duplicate server installation just to use for checking if an update is available. Luckily the Valheim server install is only 1GB so having to keep a copy isn't a huge issue (compared to something like Ark with a 12GB server). If keeping a staging server to check for updates ends up being required then it could be good to expose a mount point for it so that people running multiple Valheim containers on the same machine can all share the same staging server.

Overall this should be pretty doable and if implemented well could allow for checking for updates more frequently to reduce the chance of trying to log on to an out of date server.

> In addition to this it would be great if it could do a check before it shuts down to see if anyone is currently in the server and wait till user count = 0 before shutting down and updating.

This does sound like a nice feature, but I probably won't add it for the MVP at least. I am planning on exposing some of the information provided through [Steam queries](https://developer.valvesoftware.com/wiki/Server_queries) which includes the number of players online. There  isn't any existing work on providing this in Rust through, so I was planning on working through implementing the protocol myself to make it easy to add in here.

Ideally this information could be used to provide something like an `odin status` command that can list if the server is online, the world name, and how many players are online. A nicer solution would be if there was something like RCON to allow for interacting with Valheim, but I don't know of any way of interacting with a Valheim server remotely like that yet.

**Comment by neoKushan** [comment]
Appreciate you looking into this. I'm happy to help/test where I can, I don't know any rust but I'm a fairly experienced dev and not afraid to get my hands dirty.

**Comment by LovecraftianHorror** [comment]
Hello again everyone! I figured I would try to get all my thoughts down in writing to force me to organize everything I've found.

_Note: all paths are based on the path from being the steam user in the standard server install. i.e. from within_

```txt
❯ docker exec -it --user steam ${VALHEIM_CONTAINER} bash
```

_Note: anything that looks enough like `json` to make you just wish it was just `json` is really Valve's basically undocumented `vdf` data format. I tried out [steamy-vdf](https://lib.rs/crates/steamy-vdf) and [vdf-serde](https://lib.rs/crates/vdf-serde) to deserialize this, but had issues with both, so I'm going to planning on just manually parsing atm_

## Exploring the options from the steam thread

tl;dr I think that manually removing the cache is the best pick

### Option 1: Use a separate "staging" install to deal with the cache not updating correctly

This seems to work and also seems to be what `thmhoag/arkserver` and `lloesche/valheim-server-docker` (I think, haven't actually verified this one) use.

The downside here is that you have to keep a separate install (not a huge deal here since the server is 1GB).

### Option 2: Magic steam command

Supposedly `steamcmd +@ShutdownOnFailedCommand 1 +@NoPromptForPassword 1 +login ${SteamLoginName} +app_info_update 1 +app_status ${SteamAppID} +quit` works fine as a way of seeing if an update is required, but requires logging in with an account instead of anonymously, and also potentially doesn't always work according to others.

### Option 3: We can remove the cache :D

Good ole Marcellus Wallace mentioned that you can remove the cache at `~/Steam/appcache/appinfo.vdf` to actually get an updated response without having a staging install!

## How to check for updates

So from poking around I believe I've found that the buildid for the current install can be found at `~/valheim/steamapps/appmanifest_896660.acf` which currently for me contains

```txt
"AppState"
{
	"appid"		"896660"
	--- 8< snip 8< ---
	"buildid"		"6246034"
	--- 8< snip 8< ---
}
```

After that the current buildid available from steam can be found by removing `~/Steam/appcache/appinfo.vdf` and then requesting the app info like so

```txt
steam@13b6d458d762:~$ rm ~/Steam/appcache/appinfo.vdf
steam@13b6d458d762:~$ ~/steamcmd/steamcmd.sh +login anonymous +app_info_print "896660" +quit
```

The end of the second command contains a vdf response with the info we want (would be nice if the response was sent to `stdout` and all the junk before was sent to `stderr` instead of sending it all to `stdout` steam!)

```txt
"896660"
{
	--- 8< snip 8< ---
	"depots"
	{
		--- 8< snip 8< ---
		"branches"
		{
			"public"
			{
				"buildid"		"6246034"
				"timeupdated"		"1613558776"
			}
			--- 8< snip 8< ---
		}
	}
}
```

So from here we can see that the server is already up to date since the app manifest and the current app info both have a buildid of `6246034`, but if they differed then we would know to update!

[Example parsing buildid from app info](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ad5b0d07ac0dafe3916e68a1c8087f4b)

[Example parsing buildid from app manifest](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e8d74a97a7c79925222cdd8be200bc7b)

## Command line setup

I was thinking something along the lines of `odin update [--check | --force]`#152 

### `odin update`

Will check if there is an update. If there is no update do nothing. If there is an update then stop the server if it was running, update the server, and start the server again if it was running before (return it to the previous state).

### `odin update --check`

Checks if there is an update and then gives some easy way to include this in scripting (through a special exit code or something). This won't actually update the server. It's also possible that `odin --dry-run update` could have this same behavior instead, but it should still error if it's used with `--force` which might make people more likely to mess up using the exit code.

### `odin update --force`

Essentially equivalent to

```bash
odin stop
odin install
odin start  # <-- but only if it was running before
```

Will take down the server and try to install updates no matter what.

## Remaining concerns

I think my only remaining concern so far is from people trying to run multiple servers off the same installation. Under a typical setup only one server will go offline for the update and all the others will continue chugging along without realizing they are out of date (and may encounter issues from this as well). Honestly the server install is only 1GB, and it would still be possible to have multiple servers run off the same install, but people doing that just need to do some extra work to manage checking for updates with one server and then restarting all the other's if there was an update. I feel like this should be relatively easy to manage still with the command line options provided above.

## Closing

Any extra thoughts from anyone? Unless something seems game-changing then I should be free to implement these changes over the next couple days.

**Comment by mbround18** [comment]
@LovecraftianHorror Excellent breakdown! My suggestions are to 

- For the update checker, I like the idea of nuking the app info when we update. There are also another set of variables we could try, thoughts?

  ```sh
   ../steamcmd/steamcmd.sh +@bCSForceNoCache 1 +@ShutdownOnFailedCommand 1 +@NoPromptForPassword 1 +login anonymous +force_install_dir /home/steam/valheim  +app_info_update 1 +app_status "896660" +quit 
  ```

- You are correct in the event that the two github repos you linked and for other games they use a staging area and rsync over the files. This can be cumbersome as the storage utilization can double with a staging dir and my suggestion is if possible to skip staging.  
- As a fallback we can always keep https://docs.rs/steamworks/0.6.1/steamworks/ in mind for a potential library to utilize. 


To address your concern, we do have multiple people using this docker image with multiple servers. Some of them reuse the server binaries and run multiple servers off of them which can cause issues. My suggestion is to just make sure we have flags available for them to toggle for auto updates.

**Comment by LovecraftianHorror** [comment]
> ` ../steamcmd/steamcmd.sh +@bCSForceNoCache 1 +@ShutdownOnFailedCommand 1 +@NoPromptForPassword 1 +login anonymous +force_install_dir /home/steam/valheim  +app_info_update 1 +app_status "896660" +quit`

Oh I didn't know about that, if we could just use one command then that would be simpler. Do you know what the output looks like when there is an update available? I would assume either the `install state` or `update state` value would change, but I'm not sure what to.

> as a fallback we can always keep https://docs.rs/steamworks/0.6.1/steamworks/ in mind for a potential library to utilize.

It would be nice if we could handle all of the `steamcmd` stuff in rust if possible. I've poked through the docs a bit there, but couldn't find an equivalent to what we're trying to do here yet (granted I didn't look for too long). I think we can kick the can on this a bit if we can get stuff with steamcmd working in the meantime.

**Comment by mbround18** [comment]
I did an a lot of run around the Rosie on this :P we could diff these two commands to check if there is a version update :P

![image](https://user-images.githubusercontent.com/12646562/108919779-d620b280-75f0-11eb-9014-b3693bfb4050.png)

```sh
 ../steamcmd/steamcmd.sh +@bCSForceNoCache 1 +@ShutdownOnFailedCommand 1 +@NoPromptForPassword 1 +login anonymous +app_info_update 1 +app_info_print "896660"  +quit | grep "buildid"
```


```sh
cat steamapps/appmanifest_896660.acf | grep "buildid"
```

**Comment by LovecraftianHorror** [comment]
Getting time to dig into this a bit more (getting things together to move). I'll have to double check that the command above doesn't still use the cache, but if it doesn't then that's also a good option. I have most of the work done for the PR, I'll open up a draft before I call it for the night!

**Comment by dkurzaj** [comment]
> This does sound like a nice feature, but I probably won't add it for the MVP at least. I am planning on exposing some of the information provided through [Steam queries](https://developer.valvesoftware.com/wiki/Server_queries) which includes the number of players online. There  isn't any existing work on providing this in Rust through, so I was planning on working through implementing the protocol myself to make it easy to add in here.

Actually there seem to be this implementation : https://github.com/rumblefrog/a2s-rs

**Comment by neoKushan** [comment]
> > This does sound like a nice feature, but I probably won't add it for the MVP at least. I am planning on exposing some of the information provided through [Steam queries](https://developer.valvesoftware.com/wiki/Server_queries) which includes the number of players online. There  isn't any existing work on providing this in Rust through, so I was planning on working through implementing the protocol myself to make it easy to add in here.
> 
> Hi,
> Actually there seem to be this implementation : https://github.com/rumblefrog/a2s-rs

Is that not specific to Source games (rather than Steamworks games)? Valheim is built on Unity, not source.

**Comment by LovecraftianHorror** [comment]
@neoKushan I'll have to double check, but steam has some way of accessing the player count since it reports it in the direct server connection screen.

---

@mbround18 well now that there was another update I can confirm that `+@bCSForceNoCache 1` still uses the cache while removing the `appingo.vdf` forces the info to update :upside_down_face: 

```txt
steam@790109cf62e2:~/valheim$ ~/steamcmd/steamcmd.sh +@bCSForceNoCache 1 +@ShutdownOnFailedCommand 1 +@NoPromptForPassword 1 +login anonymous +app_info_update 1 +app_info_print "896660" +quit | grep 'buildid'
				"buildid"		"6246034"
				"buildid"		"6272190"
				"buildid"		"6287700"
steam@790109cf62e2:~/valheim$ rm ~/Steam/appcache/appinfo.vdf
steam@790109cf62e2:~/valheim$ ~/steamcmd/steamcmd.sh +@bCSForceNoCache 1 +@ShutdownOnFailedCommand 1 +@NoPromptForPassword 1 +login anonymous +app_info_update 1 +app_info_print "896660" +quit | grep 'buildid'
				"buildid"		"6287700"
				"buildid"		"6294553"
				"buildid"		"6293812"
```

**Comment by LovecraftianHorror** [comment]
Thanks for the heads up @dkurzaj ! I didn't find it because they don't include any key words relating to steam queries for crates.io, but it looks like it could be exactly what I needed for this! I just modified the unit test to point to my server and it looks like there's quite a bit of info

```txt
Info {
    protocol: 17,
    name: "[REDACTED]",
    map: "[REDACTED]",
    folder: "valheim",
    game: "",
    app_id: 0,
    players: 0,
    max_players: 64,
    bots: 0,
    server_type: Dedicated,
    server_os: Linux,
    visibility: true,
    vac: false,
    the_ship: None,
    version: "1.0.0.0",
    edf: 177,
    extended_server_info: ExtendedServerInfo {
        port: Some(
            2456,
        ),
        steam_id: Some(
            [REDACTED],
        ),
        keywords: Some(
            "0.146.8",
        ),
        game_id: Some(
            892970,
        ),
    },
    source_tv: None,
}
```
